File: frontend/src/lib/api/publicacao.ts
// O link público de acompanhamento de um quadro.
// Espelha backend/internal/adapter/http/dto/publicacao.go
//
// São duas conversas diferentes com a API, e o arquivo as mantém à vista uma da
// outra de propósito: as três primeiras são do DONO e viajam com a sessão; a
// última é a de quem chegou pelo link e não tem conta nenhuma.

import { apiDelete, apiGet, BASE_URL, ApiError } from './client';
import type { Cor, Fundo } from './boards';

/** O estado do link público, como o dono o vê. */
export interface Publicacao {
	publicado: boolean;
	/** A URL para copiar e mandar. Ausente quando o quadro não está publicado. */
	url?: string;
	criadoEm?: string;
}

/** Um rótulo no card público: já resolvido em nome e cor, sem id para cruzar. */
export interface EtiquetaPublica {
	nome: string;
	cor: Cor;
}

/**
 * Um card como quem acompanha de fora o enxerga.
 *
 * Sem responsáveis, sem comentários, sem anexos e sem id — e isso não é um
 * recorte da tela, é o que a API devolve. Ver ucboard.QuadroPublico no backend
 * para o motivo de cada ausência.
 */
export interface CardPublico {
	titulo: string;
	descricao: string;
	cor: Cor | '';
	prazo: string | null;
	vencido: boolean;
	etiquetas: EtiquetaPublica[];
	checklist: { concluidos: number; total: number };
}

export interface ColunaPublica {
	titulo: string;
	cor: Cor | '';
	cards: CardPublico[];
}

export interface QuadroPublico {
	titulo: string;
	fundo: Fundo;
	atualizadoEm: string;
	colunas: ColunaPublica[];
}

export function obterPublicacao(boardId: string): Promise<Publicacao> {
	return apiGet<Publicacao>(`/boards/${boardId}/publicacao`);
}

/**
 * publicar liga o link e devolve a URL.
 *
 * PUT, e não POST: repetir devolve o MESMO link em vez de criar um segundo.
 * Sem essa garantia, abrir a tela de compartilhamento duas vezes invalidaria em
 * silêncio o endereço já enviado às pessoas.
 */
export async function publicar(boardId: string): Promise<Publicacao> {
	const resposta = await fetch(`${BASE_URL}/boards/${boardId}/publicacao`, {
		method: 'PUT',
		credentials: 'include'
	});
	if (!resposta.ok) {
		throw new ApiError(resposta.status, 'não foi possível publicar o quadro');
	}
	return resposta.json() as Promise<Publicacao>;
}

/** despublicar derruba o link na hora. Religar depois gera um endereço NOVO. */
export function despublicar(boardId: string): Promise<void> {
	return apiDelete(`/boards/${boardId}/publicacao`);
}

/**
 * verQuadroPublico busca o quadro pelo token do link, SEM mandar cookie.
 *
 * `credentials: 'omit'` é deliberado, e é a única chamada do aplicativo que o
 * usa. A rota não olha sessão nenhuma, então mandar o cookie de quem por acaso
 * está logado só exporia a credencial dele numa requisição que não precisa
 * dela — inclusive quando a página é aberta a partir de um link recebido de
 * terceiros.
 */
export async function verQuadroPublico(token: string): Promise<QuadroPublico> {
	const resposta = await fetch(`${BASE_URL}/publico/${encodeURIComponent(token)}`, {
		credentials: 'omit'
	});
	if (!resposta.ok) {
		throw new ApiError(resposta.status, 'link inválido ou desativado');
	}
	return resposta.json() as Promise<QuadroPublico>;
}

File: backend/internal/adapter/http/dto/publicacao.go
package dto

import "time"

// PublicacaoResponse é o estado do link público de um quadro, como o dono o vê.
//
// A URL vem montada pelo servidor, e não o token cru: o link é para copiar e
// mandar, e montá-lo no cliente espalharia por lá o conhecimento de onde a
// página pública mora. É o mesmo caminho do link de convite.
type PublicacaoResponse struct {
	Publicado bool   `json:"publicado"`
	URL       string `json:"url,omitempty"`
	// CriadoEm é nulo quando o quadro não está publicado. Vai junto para a tela
	// poder dizer desde quando o link existe — "publicado" sem data não deixa
	// ninguém perceber um link esquecido aberto há meses.
	CriadoEm *time.Time `json:"criadoEm,omitempty"`
}

// QuadroPublicoResponse é o quadro servido a quem chega pelo link, sem sessão.
//
// Espelha ucboard.QuadroPublico — inclusive nas ausências, que são o motivo de
// ele existir: sem responsáveis, sem comentários, sem anexos, sem histórico,
// sem membros e sem ids. Ver o comentário daquele tipo para o porquê de cada
// uma.
type QuadroPublicoResponse struct {
	Titulo       string                  `json:"titulo"`
	Fundo        string                  `json:"fundo"`
	AtualizadoEm time.Time               `json:"atualizadoEm"`
	Colunas      []ColunaPublicaResponse `json:"colunas"`
}

// ColunaPublicaResponse é uma etapa do fluxo com os cards dela, em ordem.
type ColunaPublicaResponse struct {
	Titulo string                `json:"titulo"`
	Cor    string                `json:"cor"`
	Cards  []CardPublicoResponse `json:"cards"`
}

// CardPublicoResponse é uma tarefa como quem acompanha de fora a enxerga.
type CardPublicoResponse struct {
	Titulo    string     `json:"titulo"`
	Descricao string     `json:"descricao"`
	Cor       string     `json:"cor"`
	Prazo     *time.Time `json:"prazo"`
	Vencido   bool       `json:"vencido"`
	// As etiquetas vêm resolvidas em nome e cor: não há lista de etiquetas do
	// quadro para cruzar, porque não há id nenhum nesta resposta.
	Etiquetas []EtiquetaPublicaResponse `json:"etiquetas"`
	Checklist ProgressoResponse         `json:"checklist"`
}

// EtiquetaPublicaResponse é o rótulo de um card: só o que se lê.
type EtiquetaPublicaResponse struct {
	Nome string `json:"nome"`
	Cor  string `json:"cor"`
}
